probables/blooms/basebloom.py

```diff
@@ -230,73 +235,81 @@ def __load(
         hash_function: Union[HashFuncT, None] = None,
     ) -> None:
         """load the Bloom Filter from file"""
-        # read in the needed information, and then call _set_optimized_params
-        # to set everything correctly
         if not isinstance(file, (IOBase, mmap)):
             file = Path(file)
             with MMap(file) as filepointer:
                 self.__load(blm_type, filepointer, hash_function)
         else:
-            offset = self.__class__.HEADER_STRUCT.size
+            offset = self.__HEADER_STRUCT.size
             file.seek(offset * -1, os.SEEK_END)
-            fpr = self._parse_footer(self.__class__.HEADER_STRUCT, file.read(offset))
-            vals = self._set_optimized_params(self.__est_elements, fpr, hash_function)
-            self.__hash_func = vals[0]  # type: ignore
-            self.__fpr = vals[1]
-            self.__number_hashes = vals[2]
-            self.__num_bits = vals[3]
-            if blm_type in ["regular", "reg-ondisk"]:
-                self.__bloom_length = int(math.ceil(self.__num_bits / 8.0))
-            else:
-                self.__bloom_length = self.number_bits
+            self._parse_footer_set(self.__HEADER_STRUCT, file.read(offset), hash_function)
+            self._set_bloom_length()
             # now read in the bit array!
             file.seek(0, os.SEEK_SET)
```

Comment:
This branch is only compatible to `mmap`, because it requires the object class to simultaneously implement a `ByteString` interface and `IOBase` (for seeking). The proposal is to eliminate `IOBase` interface from this branch. Or maybe eliminate `ByteString`interface from this branch, if we are concerned about impls of python without `mmap`.

Reply:
```patch
From 552206ce840ff92f09bac2f34bb0728d54f56022 Mon Sep 17 00:00:00 2001
From: KOLANICH <KOLANICH@users.noreply.github.com>
Date: Fri, 7 Jan 2022 14:44:36 +0300
Subject: [PATCH] Made parsing of bloom filters DRYer.

---
 probables/blooms/basebloom.py | 18 +++++++-----------
 probables/blooms/bloom.py     | 10 ++++------
 2 files changed, 11 insertions(+), 17 deletions(-)

diff --git a/probables/blooms/basebloom.py b/probables/blooms/basebloom.py
index b66aa9e..5e07d3b 100644
--- a/probables/blooms/basebloom.py
+++ b/probables/blooms/basebloom.py
@@ -71,7 +71,7 @@ class BaseBloom(object):
 
         if is_valid_file(filepath):
             assert filepath is not None
-            self.__load(blm_type, filepath, hash_function)
+            self._load(filepath, hash_function)
         elif is_hex_string(hex_string):
             assert hex_string is not None
             self._load_hex(hex_string, hash_function)
@@ -228,24 +228,20 @@ class BaseBloom(object):
     __HEADER_STRUCT_BE = Struct(">" + __HEADER_STRUCT_FORMAT)
     __FPR_STRUCT = Struct("f")
 
-    def __load(
+    def _load(
         self,
-        blm_type: str,
-        file: Union[Path, str, IOBase, mmap],
+        file: Union[Path, str, IOBase, mmap, ByteString],
         hash_function: Union[HashFuncT, None] = None,
     ) -> None:
         """load the Bloom Filter from file"""
-        if not isinstance(file, (IOBase, mmap)):
+        if not isinstance(file, (IOBase, mmap, ByteString)):
             file = Path(file)
             with MMap(file) as filepointer:
-                self.__load(blm_type, filepointer, hash_function)
+                self._load(filepointer, hash_function)
         else:
-            offset = self.__HEADER_STRUCT.size
-            file.seek(offset * -1, os.SEEK_END)
-            self._parse_footer_set(self.__HEADER_STRUCT, file.read(offset), hash_function)
+            offset = len(file) - self.__HEADER_STRUCT.size
+            self._parse_footer_set(self.__HEADER_STRUCT, file[offset:], hash_function)
             self._set_bloom_length()
-            # now read in the bit array!
-            file.seek(0, os.SEEK_SET)
             self._parse_bloom_array(file)  # type: ignore
 
     @classmethod
diff --git a/probables/blooms/bloom.py b/probables/blooms/bloom.py
index 2f3d5ed..cf692a7 100644
--- a/probables/blooms/bloom.py
+++ b/probables/blooms/bloom.py
@@ -130,9 +130,7 @@ class BloomFilter(BaseBloom):
         offset = cls.__HEADER_STRUCT.size
         est_elms, _, fpr, _, _, _ = cls._parse_footer(cls.__HEADER_STRUCT, bytes(b[-offset:]))
         blm = BloomFilter(est_elements=est_elms, false_positive_rate=fpr, hash_function=hash_function)
-        blm._parse_footer_set(cls.__HEADER_STRUCT, bytes(b[-offset:]))
-        blm._set_bloom_length()
-        blm._parse_bloom_array(b)
+        blm._load(b, hash_function=blm.hash_function)
         return blm
 
     def __str__(self) -> str:
@@ -313,11 +311,11 @@ class BloomFilterOnDisk(BaseBloom):
                 a.tofile(filepointer)
                 filepointer.write(self.CNT_FOOTER_STUCT.pack(est_elements, 0, false_positive_rate))
                 filepointer.flush()
-            self.__load(filepath, hash_function)
+            self.load(filepath, hash_function)
         elif is_hex_string(hex_string):
             self._load_hex(hex_string, hash_function)
         elif is_valid_file(filepath):
-            self.__load(filepath, hash_function)
+            self.load(filepath, hash_function)
         else:
             msg = "Insufecient parameters to set up the On Disk Bloom Filter"
             raise InitializationError(msg)
@@ -350,7 +348,7 @@ class BloomFilterOnDisk(BaseBloom):
             self.__file_pointer.close()
             self.__file_pointer = None
 
-    def __load(self, filepath: Union[str, Path], hash_function: Union[HashFuncT, None] = None):
+    def load(self, filepath: Union[str, Path], hash_function: Union[HashFuncT, None] = None):
         """load the Bloom Filter on disk"""
         # read the file, set the optimal params
         # mmap everything
-- 
2.32.0
```

Reply:
I can remove the need for seek in the basebloom.py code to fully support bytestring, etc. Simply removing the os.seek and replacing it with [] notation should resolve that issue.

For the BloomFilterOnDisk, I am not too concerned about users who may not be able to use it. 

I will push the change to remove the `os.seek()` interface from the base bloom filter before merging this PR.